unit_tests/test_keystone_saml_mellon_handlers.py

```diff
@@ -0,0 +1,151 @@
+            'when': {
+                'render_config': (
+                    'endpoint.keystone-fid-service-provider.joined',
+                    'config.complete',
+                    'keystone-data.complete',),
+                'config_changed': (
+                    'endpoint.keystone-fid-service-provider.joined',),
+                'keystone_data_changed': (
+                    'endpoint.keystone-fid-service-provider.joined',),
+                'configure_websso': (
+                    'endpoint.websso-fid-service-provider.joined',
+                    'config.complete',
+                    'keystone-data.complete',
+                    'config.rendered',),
+            },
+            'when_not': {
+                'config_changed': ('config.complete',),
+                'keystone_departed': (
+                    'endpoint.keystone-fid-service-provider.joined',),
+                'keystone_data_changed': ('keystone-data.complete',),
+                'render_config': ('config.rendered',),
+                'assess_status': ('always.run',),
+            },
+        }
+        # test that the hooks were registered via the
+        # reactive.keystone_saml_mellon_handlers
+        self.registered_hooks_test_helper(handlers, hook_set, defaults)
+
+
+class TestKeystoneSAMLMellonHandlers(test_utils.PatchHelper):
+
+    def setUp(self):
+        super().setUp()
+        self.patch_release(
+            keystone_saml_mellon.KeystoneSAMLMellonCharm.release)
+        self.keystone_saml_mellon_charm = mock.MagicMock()
+        self.patch_object(handlers.charm, 'provide_charm_instance',
+                          new=mock.MagicMock())
+        self.provide_charm_instance().__enter__.return_value = (
+            self.keystone_saml_mellon_charm)
+        self.provide_charm_instance().__exit__.return_value = None
+
+        self.patch_object(handlers, 'flags')
+
+        self.uuid = 'uuid-uuid'
+        self.patch_object(handlers.uuid, 'uuid4')
+        self.uuid4.return_value = self.uuid
+
+        self.patch_object(handlers, 'unitdata',
+                          new=mock.MagicMock())
+        self.kv = mock.MagicMock()
+        self.unitdata.kv.return_value = self.kv
+
+        self.patch_object(handlers, 'endpoint_from_flag',
+                          new=mock.MagicMock())
+        self.endpoint = mock.MagicMock()
+        self.endpoint_from_flag.return_value = self.endpoint
+
+        self.protocol_name = "mapped"
+        self.remote_id_attribute = "https://samltest.id"
+        self.idp_name = "samltest"
+        self.user_facing_name = "samltest.id"
+        self.keystone_saml_mellon_charm.options.protocol_name = (
+            self.protocol_name)
+        self.keystone_saml_mellon_charm.options.remote_id_attribute = (
+            self.remote_id_attribute)
+        self.keystone_saml_mellon_charm.options.idp_name = self.idp_name
+        self.keystone_saml_mellon_charm.options.user_facing_name = (
+            self.user_facing_name)
```

Comment:
There are enough mocks here that I'm concerned that this is a functional testing framework masquerading as a unit testing framework. I'd rather see more specific, strategic mocks to test small chunks of code, with functional tests covered by zaza or somet other framework that can actually do functional tests.

Of course, charms are really, really hard to unit test without doing something like this, and unit testing coverage is useful. I'm not sure that this comment is useful beyond just grousing. :-/

Reply:
Agreed. For the reactive charms the line is blurry what bits are fully mocked out and what bits are under test. Do you have a specific example or just the general case? Not sure I can fix the general case. But I am happy to look at specific ones.

Reply:
I was just grousing in general. What's here is fine, and in keeping with what we've done elsewhere.